### PR TITLE
Calendar Widget

### LINKED ISSUE #1: Calendar Widget
@smhigley commented on [Thu Jan 19 2017](https://github.com/dojo/widget-core/issues/264)

**Enhancement**

Add an accessible calendar widget based on this mock: http://codepen.io/smhigley/pen/wodZbj

Related to #85.

### PR DESCRIPTION
**Pull request by smhigley**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

Resolves #1 

This PR adds a standalone calendar widget. It doesn't include text input and dropdown, which will be split out to a separate datepicker widget as per the issue discussion. The calendar allows all text (months, weeks, and a11y messages) to be passed in via properties for i18n, defaulting to English. It's controlled via `year`, `month`, and `selectedDate` properties for the displayed year/month, and the user-selected date. Initial API is:

```javascript
customDateCell?: any; // widget constructor that extends CalendarCell
labels?: CalendarMessages; // a11y labels
month?: number;
monthNames?: { short: string; long: string; }[];
selectedDate?: Date;
weekdayNames?: { short: string; long: string; }[];
year?: number;
renderMonthLabel?(month: number, year: number): string;
renderWeekdayCell?(day: { short: string; long: string; }): DNode;
onMonthChange?(month: number): void;
onYearChange?(year: number): void;
onDateSelect?(date: Date): void;
```

### CODE REVIEW

**Comment by tomdye** [comment]
Haven't got time to do a full review right now but can you make sure you pass the `theme` to each sub widget you create please.

File: src/calendar/Calendar.ts [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+	weekdayNames?: { short: string; long: string; }[];
+	year?: number;
+	renderMonthLabel?(month: number, year: number): string;
+	renderWeekdayCell?(day: { short: string; long: string; }): DNode;
+	onMonthChange?(month: number): void;
+	onYearChange?(year: number): void;
+	onDateSelect?(date: Date): void;
+};
+
+const TODAY = new Date();
```
Comment (agubler) [inline src/calendar/Calendar.ts]: What happens if you don't reload the module before the next day?

File: src/calendar/Calendar.ts [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+const DEFAULT_MONTHS = [
```
Comment (agubler) [inline src/calendar/Calendar.ts]: Should these be in an `i18n` bundle? As with the default weekdays and default labels.

File: src/calendar/Calendar.ts (hunk 1 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: Once the calendar is open, I don't think it should change months or years without user interaction. `TODAY` only affects which month shows up if one isn't passed with properties.

Comment (agubler) [inline src/calendar/Calendar.ts]: But that's calculated when the module is loaded, not when the calendar is opened?

Comment (smhigley) [inline src/calendar/Calendar.ts]: Should it get updated in `render()`, then? It seems like it would add a fairly frequent extra function that ideally shouldn't be needed. Another option is Calendar could require `month` and `year`, and leave it up to the parent widget to default to `TODAY`

File: src/calendar/Calendar.ts (hunk 2 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: That's fine with me! Should it go right in the Calendar widget?

Comment (mwistrand) [inline src/calendar/Calendar.ts]: Eventually at least the `DEFAULT_LABELS` should be in an `i18n` bundle, but that is true for more than just this widget. Adding the months to a bundle is trickier, since this functionality is already provided by `@dojo/i18n/date`. However, those methods require CLDR data that applications must supply themselves. So do we treat these as bundles messages, or include the necessary CLDR data here and use `@dojo/i18n/date`? This is probably a larger question that deserves its own issue, and might require a resolution to dojo/i18n#82.

Comment (agubler) [inline src/calendar/Calendar.ts]: Could you include minimal CLDR data for these (or a copy of the minimal) that gets used unless the an application has provided it's own CLDR data?

Agreed on these being a shared concern regardless of the i18n approach.

File: src/calendar/MonthPicker.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+		const { monthNames, renderMonthLabel } = this.properties;
+		return renderMonthLabel ? renderMonthLabel(month, year) : monthNames[month].long + ' ' + year;
+	}
+
+	private _renderMonthRadios() {
+		const { month, theme = {} } = this.properties;
+
+		return this.properties.monthNames.map((monthName, i) => w(Radio, {
+			key: this._radiosName + i,
+			overrideClasses: { root: css.monthRadio, input: css.monthRadioInput, checked: css.monthRadioChecked },
```
Comment (tomdye) [inline src/calendar/MonthPicker.ts]: Can you update this to `extraClasses` please?

File: src/calendar/Calendar.ts [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+	year?: number;
+	renderMonthLabel?(month: number, year: number): string;
+	renderWeekdayCell?(day: { short: string; long: string; }): DNode;
+	onMonthChange?(month: number): void;
+	onYearChange?(year: number): void;
+	onDateSelect?(date: Date): void;
+};
+
+const DEFAULT_MONTHS = [
+	{short: 'Jan', long: 'January'},
```
Comment (tomdye) [inline src/calendar/Calendar.ts]: Should these be moved to `nls` ?

File: src/calendar/MonthPicker.ts [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+			checked: i === month,
+			label: {
+				content: `<abbr title="${monthName.long}">${monthName.short}</abbr>`,
```
Comment (tomdye) [inline src/calendar/MonthPicker.ts]: should this not be separate vdom call?

File: src/calendar/Calendar.ts [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+				this._goToDate(monthLengh);
+				break;
+			case Keys.Enter:
+			case Keys.Space:
+				const { onDateSelect } = this.properties;
+				onDateSelect && onDateSelect(new Date(year, month, this._focusedDay));
+		}
+	}
+
+	private _onMonthIncrement(direction: 'up' | 'down') {
```
Comment (smhigley) [inline src/calendar/Calendar.ts]: `onMonthCrement` 😄 
I'm really bad at naming things. I'd be happy to change it if you have a better name

File: src/calendar/MonthPicker.ts (hunk 5 above)
Comment (smhigley) [inline src/calendar/MonthPicker.ts]: That's an interesting question. It's currently typed as `content: string` in form widget properties, and I'm not sure I'd want to encourage adding DNodes to label content. It could be changed, though.

File: src/calendar/Calendar.ts (hunk 4 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: Ant mentioned that too, and unfortunately that thread got collapsed. As @mwistrand  said, there's still some discussion needed around CLDR data in i18n (e.g. https://github.com/dojo/i18n/issues/82)

File: src/calendar/Calendar.ts (hunk 6 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: Also, in math at least, an increment can be positive or negative. But mostly I couldn't think of a better name :P

Comment (devpaul) [inline src/calendar/Calendar.ts]: I just happened to see this in dojo notifications and that it involved math so it piqued my curiosity... but, if you split this into two methods `_onMonthIncrement` and `_onMonthDecrement` you'd have fewer conditional blocks and wouldn't need to pass a direction.

Comment (smhigley) [inline src/calendar/Calendar.ts]: Hm, I think there would actually be more conditional blocks. Two functions = two if/else blocks vs. one if/else if/else block. It would mean no direction, though. I'm ambivalent; I actually started with separate functions and merged them b/c they shared a bunch of logic. The function has simplified since then though, so maybe it's time to split again.

Comment (smhigley) [inline src/calendar/Calendar.ts]: I'm good with that. Originally there was more repetitive logic involved so it made sense to only have one function. Now I think separating it might be best, and I like getting rid of `direction`

File: src/calendar/Calendar.ts [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+	private _onDateClick(date: number, disabled: boolean) {
+		const { onDateSelect } = this.properties;
+		let {
+			month,
+			year
+		} = this._getMonthYear();
+
+		if (disabled && date < 15) {
+			({ month, year } = this._onMonthIncrement());
```
Comment (agubler) [inline src/calendar/Calendar.ts]: do you need the wrapping brackets?

File: src/calendar/Calendar.ts [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+			month,
+			year
+		} = this._getMonthYear();
+		const { theme = {} } = this.properties;
+
+		const currentMonthLength = this._getMonthLength(month, year);
+		const previousMonthLength = this._getMonthLength(month - 1, year);
+		const initialWeekday = new Date(year, month, 1).getDay();
+
+		let dayIndex = 0,
```
Comment (agubler) [inline src/calendar/Calendar.ts]: should have a `let` per variable (plus the formatting is a bit funky!)

File: src/calendar/Calendar.ts [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+				dayIndex++;
+
+				// set isSelectedDay if the dates match
+				if (isCurrentMonth && selectedDate && new Date(year, month, date).toDateString() === selectedDate.toDateString()) {
+					isSelectedDay = true;
+				} else {
+					isSelectedDay = false;
+				}
+
+				days.push(w('date-cell', <CalendarCellProperties> {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: This should be `w<DateCell>('date-cell', ...)` now

File: src/calendar/Calendar.ts [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+
+		return weeks;
+	}
+
+	private _renderWeekdayCell(day: { short: string; long: string; }): DNode {
+		const { renderWeekdayCell } = this.properties;
+		return renderWeekdayCell ? renderWeekdayCell(day) : v('abbr', { title: day.long }, [ day.short ]);
+	}
+
+	render() {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: should be `protected render(): DNode {`

File: src/calendar/CalendarCell.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		const { date, disabled = false, onClick } = this.properties;
+		onClick && onClick(date, disabled);
+	}
+
+	private _onKeyDown(event: KeyboardEvent) {
+		const { onKeyDown } = this.properties;
+		onKeyDown && onKeyDown(event);
+	}
+
+	formatDate(date: number): DNode {
```
Comment (agubler) [inline src/calendar/CalendarCell.ts]: `public`, `protected` or `private`?

File: src/calendar/CalendarCell.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+		return v('span', [ String(date) ]);
+	}
+
+	render() {
```
Comment (agubler) [inline src/calendar/CalendarCell.ts]: `protected render(): DNode {`

File: src/calendar/MonthPicker.ts [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+		if (
+			event.which === Keys.Escape ||
+			event.which === Keys.Enter ||
+			event.which === Keys.Space
+		) {
+			this._closePopup();
+		}
+	}
+
+	private _onRadioChange(event: Event) {
```
Comment (agubler) [inline src/calendar/MonthPicker.ts]: I think that there is a `TargetedEventObject` in `widget-core` that you could use like `TargetedEventObject<HTMLInputElement>`

File: src/calendar/MonthPicker.ts [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+				break;
+			case Keys.Left:
+				this._decreaseYear();
+				break;
+		}
+	}
+
+	private _renderMonthLabel(month: number, year: number) {
+		const { monthNames, renderMonthLabel } = this.properties;
+		return renderMonthLabel ? renderMonthLabel(month, year) : monthNames[month].long + ' ' + year;
```
Comment (agubler) [inline src/calendar/MonthPicker.ts]: could use template string here?

File: src/calendar/MonthPicker.ts [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+				content: `<abbr title="${monthName.long}">${monthName.short}</abbr>`,
+				before: false
+			},
+			name: this._radiosName,
+			// TODO: replace this with a method to "unfocus" components?
+			// tabIndex: open ? 0 : -1,
+			theme,
+			value: i + '',
+			onChange: this._onRadioChange,
+			onMouseUp: () => {
```
Comment (agubler) [inline src/calendar/MonthPicker.ts]: Any reason it couldn't just be `onMouseUp: this._closePopup`?

File: src/calendar/tests/unit/Calendar.ts [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+
+let widget: Harness<CalendarProperties, typeof Calendar>;
+const testDate = new Date('June 3 2017');
+
+const compareId = compareProperty((value: any) => {
+	return typeof value === 'string';
+});
+
+const expectedDateCell = function(widget: any, date: number, active: boolean) {
+	return w('date-cell', <CalendarCellProperties> {
```
Comment (agubler) [inline src/calendar/tests/unit/Calendar.ts]: Same comment as earlier, it should be `w<DateCell>(....)`

File: src/calendar/tests/unit/Calendar.ts [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+		selected: false,
+		onClick: widget.listener,
+		onFocusCalled: widget.listener,
+		onKeyDown: widget.listener
+	});
+};
+
+const expected = function(widget: any) {
+	return v('div', { classes: widget.classes(css.root) }, [
+		w(MonthPicker, <any> {
```
Comment (agubler) [inline src/calendar/tests/unit/Calendar.ts]: shouldn't need an `any` here?

File: src/calendar/Calendar.ts [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+		return lastDate.getDate();
+	}
+
+	private _getMonthYear() {
+		const {
+			month,
+			selectedDate = this._defaultDate,
+			year
+		} = this.properties;
+		return { month: typeof month === 'number' ? month : selectedDate.getMonth(), year: year ? year : selectedDate.getFullYear() };
```
Comment (agubler) [inline src/calendar/Calendar.ts]: this is a bit hard to digest

File: src/calendar/Calendar.ts [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+		}
+
+		this._focusedDay = day;
+		this._callDateFocus = true;
+		this.invalidate();
+	}
+
+	private _onDateClick(date: number, disabled: boolean) {
+		const { onDateSelect } = this.properties;
+		let {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: For examples like this do we need to have a carriage return for each variable?

```ts
let { month, year } = this._getMonthYear();
```

File: src/calendar/Calendar.ts [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+				this._goToDate(monthLengh);
+				break;
+			case Keys.Enter:
+			case Keys.Space:
+				const { onDateSelect } = this.properties;
+				onDateSelect && onDateSelect(new Date(year, month, this._focusedDay));
+		}
+	}
+
+	private _onMonthDecrement() {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: I think you recently split these out but it does seem they share quite a lot of logic 😛

File: src/calendar/Calendar.ts [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+			onMonthChange,
+			onYearChange
+		} = this.properties;
+
+		if (month === 0) {
+			onMonthChange && onMonthChange(11);
+			onYearChange && onYearChange(year - 1);
+			return { month: 11, year: year - 1 };
+		}
+		else {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: can drop the `else` here

File: src/calendar/Calendar.ts [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+			onMonthChange,
+			onYearChange
+		} = this.properties;
+
+		if (month === 11) {
+			onMonthChange && onMonthChange(0);
+			onYearChange && onYearChange(year + 1);
+			return { month: 0, year: year + 1 };
+		}
+		else {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: same here can drop the `else`

File: src/calendar/example/index.ts [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+import { ProjectorMixin } from '@dojo/widget-core/mixins/Projector';
+import { v, w } from '@dojo/widget-core/d';
+import Calendar from '../../calendar/Calendar';
+
+const AppBase = StatefulMixin(WidgetBase);
+export class App extends AppBase<WidgetProperties> {
+	render() {
+		return v('div', {}, [
+			w(Calendar, {
+				month: <number> this.state.month,
```
Comment (agubler) [inline src/calendar/example/index.ts]: do you need the casts here?

File: src/calendar/tests/functional/Calendar.ts [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+function clickDate(remote: any) {
+	const today = new Date();
+	const firstDay = new Date(today.getFullYear(), today.getMonth(), 1).getDay();
+	return remote
+		.get('http://localhost:9000/_build/common/example/?module=calendar')
+		.setFindTimeout(5000)
+		.findByCssSelector(`tbody > tr:first-child > td:nth-child(${firstDay + 1})`)
+			.click()
```
Comment (agubler) [inline src/calendar/tests/functional/Calendar.ts]: is this formatting normal?

File: src/common/tests/unit/all.ts [hunk 25]
```diff
@@ -2,7 +2,10 @@ import '@dojo/test-extras/support/loadJsdom';
 import './main';
 import './util';
 import '../../../dialog/tests/unit/Dialog';
-import '../../../combobox/tests/unit/ComboBox';
+import '../../../calendar/tests/unit/Calendar';
+import '../../../calendar/tests/unit/CalendarCell';
+import '../../../calendar/tests/unit/MonthPicker';
+// import '../../../combobox/tests/unit/ComboBox';
```
Comment (agubler) [inline src/common/tests/unit/all.ts]: should this be commented out?

File: src/button/Button.ts [hunk 26]
```diff
@@ -35,6 +35,7 @@ export interface ButtonProperties extends ThemeableProperties {
 	content?: string;
 	describedBy?: string;
 	disabled?: boolean;
+	id?: string;
```
Comment (agubler) [inline src/button/Button.ts]: Should probably have a unit test to cover this addition.

File: src/calendar/Calendar.ts [hunk 27]
```diff
@@ -0,0 +1,391 @@
...
+ * @property weekdayNames      Customize or internationalize weekday names and abbreviations
+ * @property year              Set the currently displayed year
+ * @property renderMonthLabel  Format the displayed current month and year
+ * @property renderWeekdayCell Format the weekday column headers
+ * @property onMonthChange     Function called when the month changes
+ * @property onYearChange      Function called when the year changes
+ * @property onDateSelect      Function called when the user selects a date
+ */
+export interface CalendarProperties extends ThemeableProperties {
+	customDateCell?: any;
```
Comment (agubler) [inline src/calendar/Calendar.ts]: Is this really an `any`?

File: src/calendar/Calendar.ts (hunk 7 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: It didn't work for me otherwise, and that's the syntax I see in the MDN examples: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

Comment (agubler) [inline src/calendar/Calendar.ts]: Fair, I just tried it in node... where it is does work without the brackets.

File: src/calendar/Calendar.ts (hunk 20 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: I'll be perpetually stuck in a loop of splitting them out and merging them again 😂 .
I'm really fine either way, they seem on the edge of annoying repetitiveness vs. annoying conditionals

File: src/calendar/Calendar.ts (hunk 8 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: spaces > tabs :trollface:

File: src/calendar/MonthPicker.ts (hunk 15 above)
Comment (smhigley) [inline src/calendar/MonthPicker.ts]: I just love me some extra anonymous arrow functions 😂

File: src/common/tests/unit/all.ts (hunk 25 above)
Comment (smhigley) [inline src/common/tests/unit/all.ts]: Oh wow, I have no idea why I did that

File: src/calendar/tests/functional/Calendar.ts (hunk 24 above)
Comment (smhigley) [inline src/calendar/tests/functional/Calendar.ts]: Should I indent the `.then()` blocks? I've mostly been going off formatting in other widget functional tests.

File: src/calendar/tests/unit/Calendar.ts (hunk 17 above)
Comment (smhigley) [inline src/calendar/tests/unit/Calendar.ts]: Switched it to individual properties using `compareProperty` 👍

File: src/calendar/example/index.ts (hunk 23 above)
Comment (smhigley) [inline src/calendar/example/index.ts]: not anymore!

File: src/calendar/tests/functional/Calendar.ts (hunk 24 above)
Comment (mwistrand) [inline src/calendar/tests/functional/Calendar.ts]: And the other functional tests in `@dojo/widgets` are formatted according to the [Intern example](http://theintern.github.io/intern/#writing-functional-test), which indents the operations that pertain to a specific element.

File: src/calendar/Calendar.ts [hunk 28]
```diff
@@ -0,0 +1,389 @@
...
+
+				// set isSelectedDay if the dates match
+				if (isCurrentMonth && selectedDate && new Date(year, month, date).toDateString() === selectedDate.toDateString()) {
+					isSelectedDay = true;
+				} else {
+					isSelectedDay = false;
+				}
+
+				days.push(w<CalendarCell>('date-cell', {
+					key: date + '-' + (isCurrentMonth ? month : 'inactive'),
```
Comment (bitpshr) [inline src/calendar/Calendar.ts]: May be a nice place for a template string.

File: src/calendar/CalendarCell.ts [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+export default class CalendarCell extends CalendarCellBase<CalendarCellProperties> {
+	protected onElementCreated(element: HTMLElement, key: string) {
+		const { callFocus, onFocusCalled } = this.properties;
+		if (callFocus) {
+			element.focus();
+			onFocusCalled && onFocusCalled();
+		}
+	}
+	protected onElementUpdated(element: HTMLElement, key: string) {
+		const { callFocus, onFocusCalled } = this.properties;
```
Comment (bitpshr) [inline src/calendar/CalendarCell.ts]: You could pull out the common focus logic of `onElementCreated` and `onElementUpdated` into a private method, like `_focus`.

File: src/calendar/MonthPicker.ts [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+import uuid from '@dojo/core/uuid';
+import { Keys } from '../common/util';
+import { DNode, TypedTargetEvent } from '@dojo/widget-core/interfaces';
+import Radio from '../radio/Radio';
+import Button from '../button/Button';
+import * as css from './styles/calendar.m.css';
+import * as baseCss from '../common/styles/base.m.css';
+
+// will need to localize messages
+export interface CalendarMessages {
```
Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: I think this needs proper TypeDoc so it plays nicely with the doc generator.

File: src/calendar/MonthPicker.ts [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+	onRequestOpen?(): void;
+	onRequestClose?(): void;
+	onRequestMonthChange?(month: number): void;
+	onRequestYearChange?(year: number): void;
+};
+
+export const MonthPickerBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class MonthPicker extends MonthPickerBase<MonthPickerProperties> {
```
Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: Just curious, why didn't you put MonthPicker in the Calendar registry? I guess it's probably unlikely that consumers would need to customize it often.

File: src/calendar/MonthPicker.ts [hunk 32]
```diff
@@ -0,0 +1,299 @@
...
+	private _buttonId = uuid();
+	private _dialogId = uuid();
+	private _labelId = uuid();
+	private _yearSpinnerId = uuid();
+	private _radiosName = uuid();
```
Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: I'm not sure there'd be a huge performance gain, but couldn't you use a single `uuid()` with prefixes or suffixes for each of these?

File: src/calendar/MonthPicker.ts [hunk 33]
```diff
@@ -0,0 +1,299 @@
...
+
+	private _callTriggerFocus = false;
+	private _callPopupFocus = false;
+
+	// move focus when opening/closing the popup
+	protected onElementUpdated(element: HTMLElement, key: string) {
+		// TODO: find a way to call focus on button widget
```
Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: Are you still investigating this? Should this prevent the PR from landing?

File: src/calendar/MonthPicker.ts [hunk 34]
```diff
@@ -0,0 +1,299 @@
...
+		return this.properties.monthNames.map((monthName, i) => w(Radio, {
+			key: this._radiosName + i,
+			extraClasses: { root: css.monthRadio, input: css.monthRadioInput, checked: css.monthRadioChecked },
+			checked: i === month,
+			label: {
+				content: `<abbr title="${monthName.long}">${monthName.short}</abbr>`,
+				before: false
+			},
+			name: this._radiosName,
+			// TODO: replace this with a method to "unfocus" components?
```
Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: Same question about this TODO.

File: src/calendar/MonthPicker.ts (hunk 31 above)
Comment (smhigley) [inline src/calendar/MonthPicker.ts]: Hm, I guess I didn't really see it as something that people would customize. Right now everything in there apart from `render` is private, so it's not easily extendable. I wonder if I should start making more things protected over private, both in MonthPicker and in general?

File: src/calendar/MonthPicker.ts (hunk 33 above)
Comment (smhigley) [inline src/calendar/MonthPicker.ts]: The problem here (and below with the radio buttons) is it involves handling focus and focusability within other widgets. I could either modify all form widgets to be able to be focused and add `tabIndex` to properties or get a working focus manager. I'm inclined towards the latter, but if Calendar needs to wait for a focus manager it'll be here a while :P. I could update the comment to say it needs to be fixed once the focus manager lands and put in a link to the issue

File: src/calendar/Calendar.ts (hunk 20 above)
Comment (agubler) [inline src/calendar/Calendar.ts]: It's cool, lets leave them as two separate functions for now 👍

File: src/calendar/MonthPicker.ts [hunk 35]
```diff
@@ -0,0 +1,311 @@
...
+		const {
+			month,
+			year,
+			labelId = `${this._idBase}_label`,
+			labels,
+			open = false
+		} = this.properties;
+
+		return v('div', { classes: this.classes(css.header) }, [
+			// button
```
Comment (tomdye) [inline src/calendar/MonthPicker.ts]: I don't think these comments are required.

File: src/calendar/Calendar.ts [hunk 36]
```diff
@@ -0,0 +1,385 @@
...
+	year?: number;
+	renderMonthLabel?(month: number, year: number): string;
+	renderWeekdayCell?(day: { short: string; long: string; }): DNode;
+	onMonthChange?(month: number): void;
+	onYearChange?(year: number): void;
+	onDateSelect?(date: Date): void;
+};
+
+export const DEFAULT_MONTHS = [
+	{short: 'Jan', long: 'January'},
```
Comment (smhigley) [inline src/calendar/Calendar.ts]: Agreed this needs to change; per OSS conversation, will track in this issue and raise a separate PR: https://github.com/dojo/widgets/issues/213

File: src/calendar/Calendar.ts (hunk 27 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: @agubler @tomdye so far all the other widgets that have accepted custom widget registries in `properties` have typed them as `any` and used the `custom` prefix.

Comment (agubler) [inline src/calendar/Calendar.ts]: Well I would say we'd probably want to type them correctly everywhere, I think it would be `Constructor<Cell>` or `Constructor<WidgetBase>`? No? Assuming it accepts a `WidgetConstructor`?

File: src/calendar/Calendar.ts [hunk 37]
```diff
@@ -0,0 +1,385 @@
...
+	chooseMonth: 'Choose Month',
+	chooseYear: 'Choose Year',
+	previousMonth: 'Previous Month',
+	nextMonth: 'Next Month'
+};
+
+export const CalendarBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+@diffProperty('customDateCell', DiffType.REFERENCE)
```
Comment (bitpshr) [inline src/calendar/Calendar.ts]: I thought by default all object properties were diffed by reference, no? If not, we have a few bugs in other registry-based components I think.

File: src/calendar/Calendar.ts [hunk 38]
```diff
@@ -0,0 +1,385 @@
...
+				else {
+					date++;
+				}
+				dayIndex++;
+
+				// set isSelectedDay if the dates match
+				dateString = new Date(year, month, date).toDateString();
+				if (isCurrentMonth && selectedDate && dateString === selectedDate.toDateString()) {
+					isSelectedDay = true;
+				} else {
```
Comment (bitpshr) [inline src/calendar/Calendar.ts]: `else` should be on a new line

File: src/calendar/MonthPicker.ts [hunk 39]
```diff
@@ -0,0 +1,311 @@
...
+			}, [
+				v('span', { classes: this.classes().fixed(baseCss.visuallyHidden) }, [ labels.chooseMonth ])
+			]),
+			v('label', {
+				id: labelId,
+				classes: this.classes(css.currentMonthLabel),
+				'aria-live': 'polite',
+				'aria-atomic': 'false'
+			}, [ this._renderMonthLabel(month, year) ]),
+			// previous/next month buttons
```
Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: You can lose these comments probably.

File: src/calendar/Calendar.ts (hunk 27 above)
Comment (smhigley) [inline src/calendar/Calendar.ts]: `Constructor<CalendarCell>` seems to work, I'll use that 👍

File: src/calendar/Calendar.ts (hunk 37 above)
Comment (agubler) [inline src/calendar/Calendar.ts]: If it's a widget constructor, then it's actually by default ignored (because it's a function) - so to detect changes in between the custom constructors it does need the reference DiffType.

fwiw, object properties are swallow diff'd by default.

Comment (smhigley) [inline src/calendar/Calendar.ts]: @bitpshr I noticed it b/c tests weren't hitting `if (this._registry)`. Have you run across that in any other widgets that take a constructor in properties?

File: src/calendar/MonthPicker.ts (hunk 39 above)
Comment (smhigley) [inline src/calendar/MonthPicker.ts]: Hah, @tomdye said the same thing. It's a pretty large chunk of VDOM, though. I have trouble finding what I'm looking for at a glance, and I wrote it 😕. Are you sure we shouldn't have comments in there for later?

Comment (bitpshr) [inline src/calendar/MonthPicker.ts]: NO COMMENTS ALLOWED

No, if you needed them, I'd say leave them for sure.

File: src/calendar/README.md [hunk 40]
```diff
@@ -0,0 +1,91 @@
...
+### i18n Features
+- Localized versions of `CalendarMessages`, `monthNames`, and `weekdayNames` can be passed in via properties.
+
+## Example Usage
+
+*Basic Example*
+```js
+import Calendar from '@dojo/widgets/calendar/Calendar';
+
+protected render: DNode () {
```
Comment (bitpshr) [inline src/calendar/README.md]: I don't think we need to show the `import` of the module and a pseudocoded `render` function. This isn't valid code anyway, so why not just show the `w` call to keep boilerplate to a minimum? We also don't know if a user will be importing from the index module, this module, etc.

File: src/calendar/MonthPicker.ts (hunk 39 above)
Comment (smhigley) [inline src/calendar/MonthPicker.ts]: I made them a little more consistent and specific. If you still don't like them, I can burn them all with 🔥

File: src/calendar/Calendar.ts [hunk 41]
```diff
@@ -0,0 +1,389 @@
...
+	{short: 'Sun', long: 'Sunday'},
+	{short: 'Mon', long: 'Monday'},
+	{short: 'Tue', long: 'Tuesday'},
+	{short: 'Wed', long: 'Wednesday'},
+	{short: 'Thu', long: 'Thursday'},
+	{short: 'Fri', long: 'Friday'},
+	{short: 'Sat', long: 'Saturday'}
+];
+
+export const DEFAULT_LABELS: CalendarMessages = {
```
Comment (agubler) [inline src/calendar/Calendar.ts]: Are we nls'ing this? I cannot remember the outcome of this conversation.

Comment (smhigley) [inline src/calendar/Calendar.ts]: I believe the outcome was that it'll be a separate PR based on the consensus in https://github.com/dojo/widgets/issues/213
